pandas/tests/frame/test_join.py

```diff
@@ -0,0 +1,171 @@
+
+    def test_join_sort(self):
+        df1 = pd.DataFrame({'a': [20, 10, 0]}, index=[2, 1, 0])
+        df2 = pd.DataFrame({'b': [100, 200, 300]}, index=[1, 2, 3])
+
+        # default how='left'
+        result = df1.join(df2, sort=True)
+        expected = pd.DataFrame({'a': [0, 10, 20], 'b': [np.nan, 100, 200]},
+                                index=[0, 1, 2])
+        tm.assert_frame_equal(result, expected)
+
+        # how='left'
+        result = df1.join(df2, how='left', sort=True)
+        expected = pd.DataFrame({'a': [0, 10, 20], 'b': [np.nan, 100, 200]},
+                                index=[0, 1, 2])
+        tm.assert_frame_equal(result, expected)
+
+        # how='right' (already sorted)
+        result = df1.join(df2, how='right', sort=True)
+        expected = pd.DataFrame({'a': [10, 20, np.nan], 'b': [100, 200, 300]},
+                                index=[1, 2, 3])
+        tm.assert_frame_equal(result, expected)
+
+        # how='right'
+        result = df2.join(df1, how='right', sort=True)
+        expected = pd.DataFrame([[np.nan, 0], [100, 10], [200, 20]],
+                                columns=['b', 'a'], index=[0, 1, 2])
+        tm.assert_frame_equal(result, expected)
+
+        # how='inner'
+        result = df1.join(df2, how='inner', sort=True)
+        expected = pd.DataFrame({'a': [10, 20], 'b': [100, 200]},
+                                index=[1, 2])
+        tm.assert_frame_equal(result, expected)
+
+        # how='outer'
+        result = df1.join(df2, how='outer', sort=True)
+        expected = pd.DataFrame({'a': [0, 10, 20, np.nan],
+                                 'b': [np.nan, 100, 200, 300]},
+                                index=[0, 1, 2, 3])
+        tm.assert_frame_equal(result, expected)
+
+    def test_join_index(self):
+        # left / right
+
+        f = self.frame.loc[self.frame.index[:10], ['A', 'B']]
+        f2 = self.frame.loc[self.frame.index[5:], ['C', 'D']].iloc[::-1]
+
+        joined = f.join(f2)
+        tm.assert_index_equal(f.index, joined.index)
+        expected_columns = pd.Index(['A', 'B', 'C', 'D'])
+        tm.assert_index_equal(joined.columns, expected_columns)
+
+        joined = f.join(f2, how='left')
+        tm.assert_index_equal(joined.index, f.index)
+        tm.assert_index_equal(joined.columns, expected_columns)
+
+        joined = f.join(f2, how='right')
+        tm.assert_index_equal(joined.index, f2.index)
+        tm.assert_index_equal(joined.columns, expected_columns)
+
+        # inner
+
+        joined = f.join(f2, how='inner')
+        tm.assert_index_equal(joined.index, f.index[5:10])
+        tm.assert_index_equal(joined.columns, expected_columns)
+
+        # outer
+
+        joined = f.join(f2, how='outer')
+        tm.assert_index_equal(joined.index, self.frame.index.sort_values())
+        tm.assert_index_equal(joined.columns, expected_columns)
+
+        tm.assertRaisesRegexp(ValueError, 'join method', f.join, f2, how='foo')
```

Comment:
then make these a separate test (for these error cases)

``def test_join_index_errors``